### DISCUSSION TITLE
[OrchardCore.Media] non-admin user can not upload pictures

### DISCUSSION THREAD

**Discussion by scil** [OrchardCMS/OrchardCore#12211] [Q&A] [answered]
I have succussivly made admin upload pictures by ajax to  `/Admin/Media/Upload`.

but a normal user can not, getting response `<h1>You do not have access to this resource.</h1>`.

I have allocated all permissions to the role which the users belongs to 

![image](https://user-images.githubusercontent.com/2786775/185029094-d4adb726-9c5d-4dc0-a7fd-05aafef92dce.png)

If I add admin role to this normal user, then the user can upload picures.
![image](https://user-images.githubusercontent.com/2786775/185051878-49662c07-02bc-4f8e-9883-63c4b5760387.png)

----

the related action maybe is [OrchardCore.Media/Controllers/AdminController::Upload](https://github.com/OrchardCMS/OrchardCore/blob/de48dc263dcd38032fc5d18287d48894f5385a1b/src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs#L143)
```
        [HttpPost]
        [MediaSizeLimit]
        public async Task<ActionResult<object>> Upload(
            string path,
            string contentType,
            ICollection<IFormFile> files)
        {
            if (!await _authorizationService.AuthorizeAsync(User, Permissions.ManageMedia))
            {
                return Forbid();
            }

         ....
```

how could I debug it?

**Comment by scil** [comment, marked as answer]
now I found that I have to add another permission Access admin panel  with id `AccessAdminPanel`.

Is it safe to grant this permission  to all Authenticated usrers ?

is there any way to skip verifying this permission ?

**Reply by hishamco** [reply]
Normal users are not allowed to manage the media, that's why you got an error. If you need the users to upload some documents or images you can use Workflow to do this in their behalf

**Reply by scil** [reply]
normal user should be allowd to upload media

**Reply by hishamco** [reply]
If they have a manage media permission, they can do unexpected things ;)

**Reply by scil** [reply]
uploading media  should not require admin permission.  do i have be admin to upload img to github issues/dicussions?

**Comment by ns8482e** [comment]
@scil if you are giving a permission to a room within a building then you also need to give a permission to enter in the building 


So yes you need to give permission for AccessAdminPanel as Media app resides inside the admin

**Comment by hyzx86** [comment]
@scil  兄嘚，快到群里来QQ: 877196442

**Comment by scil** [comment]
a recipe to grant necessary permissions to allowe users to upload files

migration.recipe.v3.Permissions.upload.json
```
{
  "name": "",
  "displayName": "",
  "description": "",
  "author": "",
  "website": "",
  "version": "1",
  "issetuprecipe": false,
  "categories": "",
  "tags": [],
  "steps": [
      {
          "name": "roles",
          "Roles": [
              {
                  "Name": "Authenticated",
                  "Permissions": [ "AccessAdminPanel", "ManageMediaContent","ManageOwnMediaContent" ]
              }
          ]
      }
  ]
}
```


Migrations.cs
```
        public async Task<int> UpdateFrom2()
        {

            await _recipeMigrator.ExecuteAsync("migration.recipe.v3.Permissions.upload.json", this);
            return 3;
        }

```

**Comment by scil** [comment]
Now for security, I copied `src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs`  and changed the permission required for action `Upload`.

It's redundant. the best way is to change `OrchardCore.Media` , removing the admin-related permission requirement
